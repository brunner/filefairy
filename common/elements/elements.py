#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Common (non-reloadable) util methods for rendering templates."""

FAVICON_LINK = 'https://fairylab.surge.sh/favicon-32x32.png'


def _attr(attributes):
    if attributes is None:
        return ''

    items = sorted(attributes.items())
    return ' ' + ' '.join('{}="{}"'.format(k, v) for k, v in items)


def _trow(t, name, row):
    if row is None:
        return
    if name not in t or t[name] is None:
        t[name] = []
    t[name].append(row)


def cell(col=None, content=''):
    """Builds a cell object, describing a table cell.

    Args:
        col: The col object for this particular cell.
        content: The cell text.

    Returns:
        The cell object.
    """
    obj = {}
    if col:
        obj['col'] = col
    if content:
        obj['content'] = content
    return obj


def col(clazz='', colspan=''):
    """Builds a col object, describing a table column.

    Args:
        clazz: The CSS class(es) to apply to the column.
        colspan: The number of columns that this column should span.

    Returns:
        The col object.
    """
    obj = {}
    if clazz:
        obj['clazz'] = clazz
    if colspan:
        obj['colspan'] = colspan
    return obj


SITELINKS_HCOLS = [col(clazz='font-weight-bold text-dark')]
SITELINKS_BCOLS = [col(clazz='position-relative')]
SITELINKS_TABLES = [
    ('Tasks', [
        ('timer', '/gameday/', 'Gameday'),
        ('people', '/news/', 'News'),
        ('spreadsheet', '/standings/', 'Standings'),
    ]),
    ('Other', [
        ('dashboard', '/dashboard/', 'Dashboard'),
        ('home', '/', 'Home'),
    ]),
]


def anchor(url, content):
    """Builds an anchor element.

    Args:
        url: The link url.
        content: The inner HTML of the anchor element.

    Returns:
        The anchor element.
    """
    return '<a href="{}">{}</a>'.format(url, content)


def dialog(id_='', icon='', tables=None):
    """Builds a dialog object.

    Args:
        id_: The CSS id to set on the root dialog element.
        icon: The icon element to display in the dialog element header.
        tables: The optional list of table elements.

    Returns:
        The dialog object.
    """
    obj = {}
    if id_:
        obj['id'] = id_
    if icon:
        obj['icon'] = icon
    if tables:
        obj['tables'] = tables
    return obj


def icon_img(src, size, classes, id_):
    """Builds an icon element for an image resource.

    Args:
        src: The source of the image.
        size: The size of the image, in pixels.
        classes: The list of CSS classes to apply to the img.
        id_: The CSS id to apply to the img.

    Returns:
        The span element.
    """
    if id_:
        id_ = ' id="{}"'.format(id_)
    classes = ' class="{}"'.format(' '.join(classes)) if classes else ''
    tag = '<img src="{0}" width="{1}" height="{1}" border="0"{2}{3}>'
    return tag.format(src, size, id_, classes)


def icon_span(name='', attributes=None, classes=None):
    """Builds an icon element for an Iconic font span.

    Args:
        name: The name of the icon.
        attributes: The set of data attributes to apply to the span.
        classes: The list of CSS classes to apply to the span.

    Returns:
        The span element.
    """
    if classes is None:
        classes = []
    classes = ['oi', 'oi-' + name, 'absolute-icon'] + classes
    return span(classes=classes, attributes=attributes)


def menu():
    """Builds a menu dialog.

    Returns:
        The menu dialog.
    """
    img = icon_img(FAVICON_LINK, '16', ['absolute-icon', 'left'], '')
    span_ = span(classes=['d-block', 'px-4'], text='Fairylab')
    return dialog(id_='menu', icon=(img + span_), tables=sitelinks(home=True))


def pre(content):
    """Builds a pre element.

    Args:
        content: The inner HTML of the pre element.

    Returns:
        The pre element.
    """
    return '<pre>{}</pre>'.format(content)


def row(cells=None, attributes=None, clazz=''):
    """Builds a row object, describing a table row.

    Args:
        cells: The cells in the row.
        attributes: The set of data attributes to apply to the row.
        clazz: The CSS class(es) to apply to the row.

    Returns:
        The row object.
    """
    obj = {}
    if cells:
        obj['cells'] = cells
    if attributes:
        obj['attr'] = _attr(attributes)
    if clazz:
        obj['clazz'] = clazz
    return obj


def ruleset(selector='', rules=None):
    """Builds a CSS ruleset element.

    Args:
        selector: The CSS selector.
        rules. The list of rules.

    Returns:
        The ruleset element.
    """
    obj = {}
    if selector:
        obj['selector'] = selector
    if rules:
        obj['rules'] = rules
    return obj


def sitelinks(home=False):
    """Builds a list of sitelinks tables.

    Args:
        home: Whether or not to include a Home link in the sitelinks.

    Returns:
        The list of sitelinks tables.
    """
    tables = [topper('Site Links')]
    for head, rows in SITELINKS_TABLES:
        body = []
        for name, href, text in rows:
            if name == 'home' and not home:
                continue
            icon = icon_span(name=name, classes=['left', 'text-secondary'])
            span_ = span(classes=['d-block pl-4'], text=anchor(href, text))
            body.append(row(cells=[cell(content=(icon + span_))]))

        tables.append(
            table(
                clazz='border mb-3',
                hcols=SITELINKS_HCOLS,
                bcols=SITELINKS_BCOLS,
                head=[row(cells=[cell(content=head)])],
                body=body))

    return tables


def span(id_='', classes='', attributes=None, text=''):
    """Builds a span element.

    Args:
        id: The CSS id to apply to the span.
        classes: The list of CSS classes to apply to the span.
        attributes: The set of data attributes to apply to the span.
        content: The inner HTML of the span element.

    Returns:
        The span element.
    """
    attr = _attr(attributes)
    if id_:
        id_ = ' id="{}"'.format(id_)
    if classes:
        classes = ' class="{}"'.format(' '.join(classes))
    return '<span{}{}{}>{}</span>'.format(id_, classes, attr, text)


def table(clazz='',
          id_='',
          attributes=None,
          hcols=None,
          bcols=None,
          fcols=None,
          head=None,
          body=None,
          foot=None):
    """Builds a table element.

    Args:
        clazz: The optional CSS class to set on the root table element.
        id_: The optional CSS id to set on the root table element.
        attributes: The set of data attributes to apply to the table.
        hcols: The optional list of CSS classes to apply to each head column.
        bcols: The optional list of CSS classes to apply to each body column.
        fcols: The optional list of CSS classes to apply to each footer column.
        head: The optional list of head rows.
        body: The optional list of body rows.
        foot: The optional list of footer rows.

    Returns:
        The table element.
    """
    obj = {}
    if clazz:
        obj['clazz'] = clazz
    if id_:
        obj['id'] = id_
    if attributes:
        obj['attributes'] = _attr(attributes)
    if hcols:
        obj['hcols'] = hcols
    if bcols:
        obj['bcols'] = bcols
    if fcols:
        obj['fcols'] = fcols
    if head:
        obj['head'] = head
    if body:
        obj['body'] = body
    if foot:
        obj['foot'] = foot
    return obj


def tbody(obj, row):
    """Appends a body row to an existing table.

    Args:
        obj: The table.
        row: The row.
    """
    _trow(obj, 'body', row)


def topper(text):
    """Builds a topper table element.

    Args:
        text: The text of the topper table element.

    Returns:
        The topper table element.
    """
    return table(
        clazz='topper',
        bcols=[col(clazz='border-0 font-weight-bold px-0 text-secondary')],
        body=[row(cells=[cell(content=text)])])


def ts(half, minute, second):
    """Returns a string representing a livesim timestamp.

    Args:
        half: The index of the inning.
        minute: The index of the minute.
        second: The index of the second.

    Returns:
        The timestamp string.
    """
    return '{}.{}.{}'.format(half, minute, second)
