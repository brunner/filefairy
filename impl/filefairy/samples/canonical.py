#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Sample data for filefairy.py golden test."""

import os
import re
import sys

_path = os.path.dirname(os.path.abspath(__file__))
sys.path.append(re.sub(r'/core/fairylab/samples', '', _path))

from common.elements.elements import card  # noqa

_external = [
    card(
        href='/fairylab/baz/',
        title='baz',
        info='Description of baz.',
        ts='06:02:30 EDT (1985-10-26)',
        danger='disabled'),
    card(
        href='/fairylab/foo/',
        title='foo',
        info='Description of foo.',
        ts='06:02:30 EDT (1985-10-26)',
    ),
    card(
        href='/fairylab/quuz/',
        title='quuz',
        info='Description of quuz.',
        ts='06:02:30 EDT (1985-10-26)',
    ),
    card(
        href='/fairylab/qux/',
        title='qux',
        info='Description of qux.',
        ts='06:02:30 EDT (1985-10-26)',
    )
]
_internal = [
    card(
        title='bar',
        info='Description of bar.',
        ts='06:02:30 EDT (1985-10-26)',
    ),
    card(
        title='quux',
        info='Description of quux.',
        ts='06:02:30 EDT (1985-10-26)',
        danger='disabled'),
]

subtitle = ''

tmpl = 'home.html'

context = {
    'external': _external,
    'internal': _internal,
}
