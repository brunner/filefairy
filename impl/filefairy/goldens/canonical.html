<!doctype html>
<html lang="en">

<head>

<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link rel="apple-touch-icon" sizes="120x120" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="manifest" href="/site.webmanifest">
<meta name="msapplication-TileColor" content="#dfe3e7">
<meta name="theme-color" content="#dfe3e7">
<link rel="stylesheet"
      href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css"
      integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm"
      crossorigin="anonymous">
<link rel="stylesheet"
      href="https://cdnjs.cloudflare.com/ajax/libs/open-iconic/1.1.1/font/css/open-iconic-bootstrap.min.css">
<title>Fairylab | fairylab</title>

<style>
html {
  position: relative;
  min-height: 100%;
}
body {
  margin-bottom: 60px;
}
header {
  background-color: #f5f5f5;
  width: 100%;
}
footer {
  background-color: #f5f5f5;
  bottom: 0;
  height: 60px;
  line-height: 60px;
  position: absolute;
  width: 100%;
}
a.disabled{
  pointer-events: none;
  cursor: default;
}
.alert, .dropdown-toggle, .form-control {
  padding: .75rem;
}
.alert .oi {
  line-height: 24px;
  padding-right: .75rem;
}
.anchor {
  padding-top: 0px;
  margin-top: 0px;
  transition: padding-top .35s ease, margin-top .35s ease;
}
.badge-icon {
  border: 1px solid #dee2e6;
  font-weight: 400!important;
  padding: .25rem .5rem;
  white-space: normal;
  width: 100%;
}
.badge-icon[data-toggle="modal"]:hover {
  background-color: #f5f5f5;
  cursor: pointer;
}
.breadcrumb {
  background-color: inherit;
  line-height: 60px;
  max-width: 468px;
  overflow-x: auto;
}
.dropdown-toggle {
  color: #6c767c;
}
.flex-grow {
  flex-grow: 1;
}
.grayscale {
  -webkit-filter: grayscale(1) opacity(0.5);
  filter: grayscale(1) opacity(0.5);
}
.left-8p {
  left: 8px;
}
.modal-header {
  height: 56px;
}
.modal-body {
  padding-top: 0;
}
.mr-8p {
  margin-right: 8px;
}
.nav-pills {
  overflow-x: auto;
  overflow-y: hidden;
}
.navbar {
  padding: 0;
}
.nav-pills .nav-item {
  margin: 0 .25rem;
  min-width: 32px;
  text-align: center;
}
.nav-pills .nav-link {
  padding: .25rem 0;
}
.nav-pills .nav-link:not(.active) {
  border: 1px solid #dee2e6;
}
.pl-24p {
  padding-left: 24px;
}
.pr-24p {
  padding-right: 24px;
}
.right-8p {
  right: 8px;
}
.table, .table-sm {
  margin-bottom: 0;
  word-break: break-word;
}
.table .thead-light td {
  background-color: #f5f5f5;
}
.table-fixed {
  table-layout: fixed;
}
.table-sm td {
  border-top: 0;
  padding: 0 .3rem 0 0;
}
.table .td-sm {
  padding: .75rem .25rem;
}
.table.collapse.show {
  display: table;
}
.text-underline {
  text-decoration: underline;
}
.top-14p {
  top: 14px;
}
.w-10 {
  width: 10%;
}
.w-20 {
  width: 20%;
}
.w-40 {
  width: 40%;
}
.w-24p {
  width: 24px;
}
.w-28p {
  width: 28px;
}
.w-55p {
  width: 55px;
}
.w-65p {
  width: 65px;
}
.w-75p {
  width: 75px;
}
.w-105p {
  width: 105px;
}
.td-sm-none {
  display: none;
}
@media (min-width: 576px) {
  body.modal-open .navbar,
  body.modal-open .timestamp {
    padding-left: 8px;
  }
  .td-sm-none {
    display: table-cell;
  }
}
</style>
</head>

<body>

<!-- ^(base.header) -->
<header class="border-bottom sticky-top">
<div class="container">
<nav class="navbar navbar-light border-bottom" aria-label="nav">
<ol class="breadcrumb flex-nowrap text-truncate m-0 p-0">
<li class="breadcrumb-item active" aria-current="page">Fairylab</li>
</ol>
</nav>

<div class="timestamp my-1">
<small class="text-muted">Last reloaded: 06:02:30 PDT (1985-10-26)</small>
</div>

</div>
</header>
<!-- $(base.header) -->

<!-- ^(base.main) -->
<main role="main" class="pb-3">

<!-- ^(home.external) -->
<div class="container anchor" id="external">

<!-- ^(common.heading) -->
<div class="row">
<div class="col-12">
<div class="alert alert-secondary mt-3 mb-0">
<span class="oi oi-globe text-secondary"></span>External Tasks
</div>
</div>
</div>
<!-- $(common.heading) -->

<!-- ^(home.row) -->
<div class="row d-flex flex-wrap">

<!-- ^(home.column) -->
<div class="col-md-6 col-lg-4 d-flex flex-column">

<!-- ^(common.card) -->
<div class="card d-flex flex-column flex-grow mt-3">

<!-- ^(common.card.body) -->
<div class="card-body d-flex flex-column">
<div class="d-flex flex-row">
<h5 class="card-title flex-grow">
<a href="/fairylab/baz/">baz</a>
</h5>
<div><span class="btn btn-danger btn-sm disabled">disabled</span></div>
</div>

<p class="card-text flex-grow">Description of baz.</p>
</div>
<!-- $(common.card.body) -->

<!-- ^(common.card.footer) -->
<div class="card-footer text-muted text-truncate">
<small>Last activity: 06:02:30 EDT (1985-10-26)</small>
</div>
<!-- $(common.card.footer) -->

</div>
<!-- $(common.card) -->

</div>
<!-- $(home.column) -->

<!-- ^(home.column) -->
<div class="col-md-6 col-lg-4 d-flex flex-column">

<!-- ^(common.card) -->
<div class="card d-flex flex-column flex-grow mt-3">

<!-- ^(common.card.body) -->
<div class="card-body d-flex flex-column">
<div class="d-flex flex-row">
<h5 class="card-title flex-grow">
<a href="/fairylab/foo/">foo</a>
</h5>
<div></div>
</div>

<p class="card-text flex-grow">Description of foo.</p>
</div>
<!-- $(common.card.body) -->

<!-- ^(common.card.footer) -->
<div class="card-footer text-muted text-truncate">
<small>Last activity: 06:02:30 EDT (1985-10-26)</small>
</div>
<!-- $(common.card.footer) -->

</div>
<!-- $(common.card) -->

</div>
<!-- $(home.column) -->

<!-- ^(home.column) -->
<div class="col-md-6 col-lg-4 d-flex flex-column">

<!-- ^(common.card) -->
<div class="card d-flex flex-column flex-grow mt-3">

<!-- ^(common.card.body) -->
<div class="card-body d-flex flex-column">
<div class="d-flex flex-row">
<h5 class="card-title flex-grow">
<a href="/fairylab/quuz/">quuz</a>
</h5>
<div></div>
</div>

<p class="card-text flex-grow">Description of quuz.</p>
</div>
<!-- $(common.card.body) -->

<!-- ^(common.card.footer) -->
<div class="card-footer text-muted text-truncate">
<small>Last activity: 06:02:30 EDT (1985-10-26)</small>
</div>
<!-- $(common.card.footer) -->

</div>
<!-- $(common.card) -->

</div>
<!-- $(home.column) -->

<!-- ^(home.column) -->
<div class="col-md-6 col-lg-4 d-flex flex-column">

<!-- ^(common.card) -->
<div class="card d-flex flex-column flex-grow mt-3">

<!-- ^(common.card.body) -->
<div class="card-body d-flex flex-column">
<div class="d-flex flex-row">
<h5 class="card-title flex-grow">
<a href="/fairylab/qux/">qux</a>
</h5>
<div></div>
</div>

<p class="card-text flex-grow">Description of qux.</p>
</div>
<!-- $(common.card.body) -->

<!-- ^(common.card.footer) -->
<div class="card-footer text-muted text-truncate">
<small>Last activity: 06:02:30 EDT (1985-10-26)</small>
</div>
<!-- $(common.card.footer) -->

</div>
<!-- $(common.card) -->

</div>
<!-- $(home.column) -->

</div>
<!-- $(home.row) -->

</div>
<!-- $(home.external) -->

<!-- ^(home.internal) -->
<div class="container anchor" id="internal">

<!-- ^(common.heading) -->
<div class="row">
<div class="col-12">
<div class="alert alert-secondary mt-3 mb-0">
<span class="oi oi-dashboard text-secondary"></span>Internal Tasks
</div>
</div>
</div>
<!-- $(common.heading) -->

<!-- ^(home.row) -->
<div class="row d-flex flex-wrap">

<!-- ^(home.column) -->
<div class="col-md-6 col-lg-4 d-flex flex-column">

<!-- ^(common.card) -->
<div class="card d-flex flex-column flex-grow mt-3">

<!-- ^(common.card.body) -->
<div class="card-body d-flex flex-column">
<div class="d-flex flex-row">
<h5 class="card-title flex-grow">
bar
</h5>
<div></div>
</div>

<p class="card-text flex-grow">Description of bar.</p>
</div>
<!-- $(common.card.body) -->

<!-- ^(common.card.footer) -->
<div class="card-footer text-muted text-truncate">
<small>Last activity: 06:02:30 EDT (1985-10-26)</small>
</div>
<!-- $(common.card.footer) -->

</div>
<!-- $(common.card) -->

</div>
<!-- $(home.column) -->

<!-- ^(home.column) -->
<div class="col-md-6 col-lg-4 d-flex flex-column">

<!-- ^(common.card) -->
<div class="card d-flex flex-column flex-grow mt-3">

<!-- ^(common.card.body) -->
<div class="card-body d-flex flex-column">
<div class="d-flex flex-row">
<h5 class="card-title flex-grow">
quux
</h5>
<div><span class="btn btn-danger btn-sm disabled">disabled</span></div>
</div>

<p class="card-text flex-grow">Description of quux.</p>
</div>
<!-- $(common.card.body) -->

<!-- ^(common.card.footer) -->
<div class="card-footer text-muted text-truncate">
<small>Last activity: 06:02:30 EDT (1985-10-26)</small>
</div>
<!-- $(common.card.footer) -->

</div>
<!-- $(common.card) -->

</div>
<!-- $(home.column) -->

</div>
<!-- $(home.row) -->

</div>
<!-- $(home.internal) -->

</main>
<!-- $(base.main) -->

<script src="https://code.jquery.com/jquery-3.2.1.slim.min.js"
        integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN"
        crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js"
        integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q"
        crossorigin="anonymous"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta.3/js/bootstrap.min.js"
        integrity="sha384-a5N7Y/aK3qNeh15eJKGWxsqtnX/wWdSZSKp+81YjTmS15nvnvxKHuzaWwXHDli+4"
        crossorigin="anonymous"></script>


</body>

</html>